lib/Sema/NameBinding.cpp

```diff
@@ -36,60 +37,539 @@
+  /// source, or it may represent a cross-import overlay that has been found and
+  /// needs to be loaded.
+  struct UnboundImport {
+    /// If this UnboundImport directly represents an ImportDecl, the ImportDecl
+    /// it represents.
+    ///
+    /// This property should only be used to control where information is added
+    /// to the AST. Don't retrieve information directly from the ImportDecl; use
+    /// the other member variables.
+    NullablePtr<ImportDecl> ID;
+
+    SourceLoc importLoc;
+    ImportOptions options;
+    StringRef privateImportFileName;
+    Located<ImportKind> importKind;
+    ModuleDecl::AccessPathTy modulePath;
+    ModuleDecl::AccessPathTy declPath;
+
+    NullablePtr<DeclAttributes> attrs;
+
+    /// Create an UnboundImport for a user-written import declaration.
+    explicit UnboundImport(ImportDecl *ID);
+
+    /// Make sure the import is not a self-import.
+    bool checkNotTautological(const SourceFile &SF);
+
+    /// Make sure the module actually loaded, and diagnose if it didn't.
+    bool checkModuleLoaded(ModuleDecl *M, SourceFile &SF);
+
+    /// Find the top-level module for this module. If \p M is not a submodule,
+    /// returns \p M. If it is a submodule, returns either the parent module of
+    /// \p M or \c nullptr if the current module is the parent (which can happen
+    /// in a mixed-language framework).
+    NullablePtr<ModuleDecl> getTopLevelModule(ModuleDecl *M, SourceFile &SF);
+
+    /// Diagnose any errors concerning the \c @_exported, \c @_implementationOnly,
+    /// \c @testable, or \c @_private attributes, including a
+    /// non-implementation-only import of a fragile library from a resilient one.
+    void validateOptions(NullablePtr<ModuleDecl> topLevelModule, SourceFile &SF);
+
+    /// Create an \c ImportedModuleDesc from the information in this
+    /// UnboundImport.
+    ImportedModuleDesc makeDesc(ModuleDecl *module) const {
+      return ImportedModuleDesc({ declPath, module }, options,
+                                privateImportFileName);
+    }
 
-namespace {  
-  class NameBinder {    
-  public:
+  private:
+    void validatePrivate(ModuleDecl *topLevelModule);
+    void validateImplementationOnly(ASTContext &ctx);
+    void validateTestable(ModuleDecl *topLevelModule);
+    void validateResilience(NullablePtr<ModuleDecl> topLevelModule,
+                            SourceFile &SF);
+
+    /// Diagnoses an inability to import \p modulePath in this situation and, if
+    /// \p attrs is provided and has an \p attrKind, invalidates the attribute and
+    /// offers a fix-it to remove it.
+    void diagnoseInvalidAttr(DeclAttrKind attrKind, DiagnosticEngine &diags,
+                             Diag<Identifier> diagID);
+  };
+
+  /// Represents an import whose options have been checked and module has been
+  /// loaded, but its scope (if any) has not been validated.
+  struct BoundImport {
```

Comment:
1. I think it is good that you separated out `BoundImport` and `UnboundImport`. This makes the separation of responsibilities clearer.
2. It is not entirely clear what "scope has not been validated" means here, and the doc comment for `validateScope` doesn't help. Moreover, there is a separate field `needsScopeValidation` defaults to `false`, which seems to contradict this (if it has not been validated, then why is the default value false?). Perhaps phrase it to something like:

```
  /// Represents an import whose options have been checked and module has been
   /// loaded, but its scope has potentially not been validated.
   ///
   /// Scope validation is needed for imports like `import func Mod.fn` where we check that `Mod`
   /// does contain `fn`.
```

?
```

Reply:
```
  /// Represents an import whose options have been checked and module has been
  /// loaded, but its scope (if it's a scoped import) has not been validated
  /// and it has not been added to \c SF.
```

Reply:
I've also expanded the doc comment on `validateScope`:

```
    /// Validate the scope of the import, if needed.
    ///
    /// A "scoped import" is an import which only covers one particular
    /// declaration, such as:
    ///
    ///     import class Foundation.NSString
    ///
    /// We validate the scope by making sure that the named declaration exists
    /// and is of the kind indicated by the keyword. This can't be done until we
    /// have bound all cross-import overlays, since a cross-import overlay could
    /// provide the declaration.
```

And I've flipped the sense of the `needsScopeValidation` so I could give it a better name and more accurate doc comment:

```
    /// If \c true, another, more specific \c BoundImport will validate the same
    /// scope as this import, so validating this one is not necessary.
    bool scopeValidatedElsewhere = true;
```